plugins/right-sizer/src/util/kubernetes.go

```diff
@@ -68,3 +83,194 @@ func Clientset() KubeClientResources {
+	return patchedResource, nil
+}
+
+// GetUnstructuredResourceFromObjectRef fffetches a resource from in-cluster, based on
+// the GroupVersionKind from the provided core.ObjectReference. An
+// ObjectReference is typically included in a Kube Event.
+func GetUnstructuredResourceFromObjectRef(kubeClientResources KubeClientResources, objectRef core.ObjectReference) (resource *unstructured.Unstructured, found bool, err error) {
+	// A GroupVersionKind is required to create the RESTMapping, which maps;
+	// converts the API version and kind to the correct capitolization and plural
+	// syntax required by the Kube API.
+	GVK := objectRef.GroupVersionKind()
+	GVKMapping, err := kubeClientResources.RESTMapper.RESTMapping(GVK.GroupKind(), GVK.Version)
+	if err != nil {
+		return nil, false, fmt.Errorf("error creating RESTMapper mapping from group-version-kind %v: %v", GVK, err)
+	}
+	getterClient := kubeClientResources.DynamicClient.Resource(GVKMapping.Resource).Namespace(objectRef.Namespace)
+	glog.V(3).Infof("starting fetch resource %s %s/%s in-cluster to get the full resource from an objectRef", objectRef.Kind, objectRef.Namespace, objectRef.Name)
+	resource, err = getterClient.Get(context.TODO(), objectRef.Name, metav1.GetOptions{})
+	if k8sErrors.IsNotFound(err) {
+		glog.V(3).Infof("resource %s %s/%s not found in-cluster while getting the full resource from an objectRef", objectRef.Kind, objectRef.Namespace, objectRef.Name)
+		return nil, false, nil
+	}
+	if err != nil {
+		glog.V(3).Infof("error fetching %s %s/%s in-cluster while getting the full resource from an objectRef: %v", objectRef.Kind, objectRef.Namespace, objectRef.Name, err)
+		return nil, false, err
+	}
+	glog.V(3).Infof("finished fetch resource %s %s/%s in-cluster to get the full resource from an objectRef", objectRef.Kind, objectRef.Namespace, objectRef.Name)
+	return resource, true, nil
+}
+
+// GetControllerFromPod accepts a typed pod object, and returns the pod-controller
+// which owns the pod.
+// E.G. an owning pod-controller might be a Kubernetes Deployment, DaemonSet,
+// or CronJob.
+func GetControllerFromPod(kubeClientResources KubeClientResources, pod *core.Pod) (*unstructured.Unstructured, error) {
+	// Convert a pod type to an unstructured one.
+	podJSON, err := json.Marshal(pod)
+	if err != nil {
+		return nil, err
+	}
+	objectAsMap := make(map[string]interface{})
+	err = json.Unmarshal(podJSON, &objectAsMap)
+	if err != nil {
+		return nil, err
+	}
+	unstructuredPod := unstructured.Unstructured{
+		Object: objectAsMap,
+	}
+
+	topController, err := fwControllerUtils.GetTopController(context.TODO(), kubeClientResources.DynamicClient, kubeClientResources.RESTMapper, unstructuredPod, nil)
+	if err != nil {
+		return nil, err
+	}
+	glog.V(2).Infof("found controller kind %q named %q for pod %s/%s", topController.GetKind(), topController.GetName(), pod.Namespace, pod.Name)
+	return &topController, nil
+}
```

Comment:
this stuff will end up in controller-utils right?

Reply:
Yes that's the eventual plan. I haven't done it so far because the functions I'm creating for right-sizer keep changing enough IMO.